### PR TITLE
Emit hidden version of a secret key in log.trace

### LINKED ISSUE #719: Logging should not expose any secret keys
```
node-2_1_37a10f21eb7b | 2020-04-02T01:15:19.329464369Z 2020-04-02 01:15:19,329 Trace [agora.node.Node] - Config is: { banman: { max_failed_requests: 10, ban_duration: 86400 }, node: { is_validator: true, min_listeners: 2, max_listeners: 10, address: "0.0.0.0", port: 4826, key_pair: { address: GA3DMXTREDC4AIUTHRFIXCKWKF7BDIXRWM2KLV74OPK2OKDM2VJ235GN, secret: 0xad53d56c28a7d573fcd7a534b3f1a2117e5156898b4a3c9322c0c520715e36364bdcf8f0442d0de2ff5f954e092c4d1cbc43ae832b336eb7e80da9a7ba45c911, seed: SAI4SRN2U6UQ32FXNYZSXA5OIO6BYTJMBFHJKX774IGS2RHQ7DOEW5SJ }, retry_delay: 1500, max_retries: 5, timeout: 1000, data_dir: ".cache" }, admin: { enabled: false, address: "127.0.0.1", port: 2827 }, network: ["http://node-0:2826", "http://node-1:3826"], dns_seeds: [], quorum: { threshold: 2, nodes: [GBUVRIIBMHKC4PE6BK7MO2O26U2NJLW4WGGWKLAVLAA2DLFZTBHHKOEK, GBJABNUCDJCIL5YJQMB5OZ7VCFPKYLMTUXM2ZKQJACT7PXL7EVOMEKNZ, GA3DMXTREDC4AIUTHRFIXCKWKF7BDIXRWM2KLV74OPK2OKDM2VJ235GN], quorums: [] }, logging: { log_level: 0 } }
```

> secret: 0xad53d56c28a7d573fcd7a534b3f1a2117e5156898b4a3c9322c0c520715e36364bdcf8f0442d0de2ff5f954e092c4d1cbc43ae832b336eb7e80da9a7ba45c911, seed: SAI4SRN2U6UQ32FXNYZSXA5OIO6BYTJMBFHJKX774IGS2RHQ7DOEW5SJ

This is a security issue. If someone finds a bug in Agora and wants to file a bug then they might provide us with a log dump. The log should not emit any private keys.

Rather than just removing its output, I can think of alternatives:
- Emit a hashed version of the key. It may help with logging in case that key appears in multiple error messages.
- Emit a shortened pretty-printed version of the key.

@Geod24 any thoughts?

### PR DESCRIPTION
**Pull request by kchulj**
Logging should not expose any secret keys. Therefore, `log.trace` now emits a hidden version of `SecretKey`, `Seed`, and `Scalar` by default (`PrintMode.Obfuscated`). Switching the argument to `PrintMode.Clear` argument to `toString` prints the original value.

`PrintMode.Obfuscated`:
```
validator: { enabled: true, key_pair: { address: GCOQEOHAUFYUAC6G22FJ3GZRNLGVCCLESEJ2AXBIJ5BJNUVTAERPLRIJ, secret: **SECRET**, seed: **SEED** }
```
`PrintMode.Clear`:
```
validator: { enabled: true, key_pair: { address: GCOQEOHAUFYUAC6G22FJ3GZRNLGVCCLESEJ2AXBIJ5BJNUVTAERPLRIJ, secret: 0xf52201b3d296424f285ca01391640951cd6a319b9d8ad6c60b4071a1e038029d94b84c9c7e3c72d9c836046c5cd87b5beb6c8e6498ea1b757207e7c22d29c5a7, seed: SCT4KKJNYLTQO4TVDPVJQZEONTVVW66YLRWAINWI3FZDY7U4JS4JJEI4 } 
```
Fixes #719

### CODE REVIEW

File: source/agora/common/crypto/ECC.d [hunk 1]
```diff
@@ -90,11 +91,29 @@ public struct Scalar
         crypto_core_ed25519_scalar_reduce(this.data[].ptr, param[].ptr);
     }
 
+    public enum PrintMode
```
Comment (AndrejMitrovic) [inline source/agora/common/crypto/ECC.d]: Needs `///` style comment. And also a description of what each mode does in the comment.

File: source/agora/common/crypto/Key.d [hunk 2]
```diff
@@ -259,6 +260,30 @@ public struct SecretKey
         this.data = args;
     }
 
+    public enum PrintMode
```
Comment (AndrejMitrovic) [inline source/agora/common/crypto/Key.d]: Also needs a comment.

File: source/agora/common/crypto/Key.d [hunk 3]
```diff
@@ -259,6 +260,30 @@ public struct SecretKey
...
+    {
+        Obfuscated,
+        Clear,
+    }
+
+    public void toString (scope void delegate(const(char)[]) @safe sink,
```
Comment (AndrejMitrovic) [inline source/agora/common/crypto/Key.d:277]: Also needs documentation.

File: source/agora/common/crypto/Key.d [hunk 4]
```diff
@@ -311,6 +336,12 @@ public struct Seed
         this.data = args;
     }
 
+    public enum PrintMode
```
Comment (AndrejMitrovic) [inline source/agora/common/crypto/Key.d]: Also documentation needed.

File: source/agora/consensus/PreImage.d [hunk 5]
```diff
@@ -75,7 +75,7 @@ public struct PreImageCache
     ***************************************************************************/
 
     public void toString (scope void delegate(scope const(char)[]) @safe sink,
-                          PrintMode mode = PrintMode.Bounds) const
+                          PrintMode mode = PrintMode.Bounds) const @safe
```
Comment (AndrejMitrovic) [inline source/agora/consensus/PreImage.d]: This belongs in its own commit.

File: source/agora/common/crypto/Key.d [hunk 6]
```diff
@@ -271,7 +296,7 @@ public struct SecretKey
 
     ***************************************************************************/
 
-    public Signature sign (scope const(ubyte)[] msg) const
+    public Signature sign (scope const(ubyte)[] msg) const nothrow @nogc
```
Comment (AndrejMitrovic) [inline source/agora/common/crypto/Key.d:331]: This should ideally be moved into its own commit. So one commit replacing the `nothrow @nogc:` with this, and a separate commit which only implements what the issue is about.

**Review by AndrejMitrovic — DISMISSED** [review]
Where are the unittests?

File: source/agora/common/crypto/ECC.d [hunk 7]
```diff
@@ -20,6 +20,7 @@ module agora.common.crypto.ECC;
 import agora.common.Hash;
```
Comment (AndrejMitrovic) [inline source/agora/common/crypto/ECC.d]: The commit message needs a description and a note of which issue it fixes. For example:

```
Emit hashed version of a secret key by default

To avoid leaking private keys into logs, we emit a shortened hashed version of the key. 
The behavior can toggled off with the `PrintMode` argument to `toString`.

Fixes #719
```

File: source/agora/common/crypto/ECC.d [hunk 8]
```diff
@@ -90,11 +91,46 @@ public struct Scalar
         crypto_core_ed25519_scalar_reduce(this.data[].ptr, param[].ptr);
     }
 
-    /// Expose `toString`
-    public void toString (scope void delegate(const(char)[]) @safe sink)
-        const @safe
+    /// Describe print modes for `toString`
+    public enum PrintMode
+    {
+        /// Default value, print hashed version
```
Comment (AndrejMitrovic) [inline source/agora/common/crypto/ECC.d]:
```suggestion
        /// Print shortened hashed version
```

The default value is in the parameter itself. And users know that the first member of an enum is the enum's `.init` value, it doesn't need to be explained here.

File: source/agora/common/crypto/ECC.d [hunk 9]
```diff
@@ -90,11 +91,46 @@ public struct Scalar
...
+            this.data[].hashFull().prettify().toString(sink);
+            break;
+
+        case PrintMode.Clear:
+            this.data.toString(sink);
+            break;
+
+        default:
+            assert(0);
+        }
```
Comment (AndrejMitrovic) [inline source/agora/common/crypto/ECC.d]:
```suggestion
        final switch (mode)
        {
        case PrintMode.Obfuscated:
            this.data[].hashFull().prettify().toString(sink);
            break;

        case PrintMode.Clear:
            this.data.toString(sink);
            break;
        }
```

Final switch is a feature in D. If you have a missing `case` it will error at compile-time, rather than having the exception being thrown at runtime.

File: source/agora/common/crypto/Key.d [hunk 10]
```diff
@@ -258,6 +260,49 @@ public struct SecretKey
...
+          sink = The sink to write the piecemeal string data to
+          mode = The `PrintMode` to use for printing the content.
+                 By default, the hashed value is printed.
+
+    ***************************************************************************/
+
+    public void toString (scope void delegate(const(char)[]) @safe sink,
+                          PrintMode mode = PrintMode.Obfuscated) const @safe
+    {
+        switch (mode)
```
Comment (AndrejMitrovic) [inline source/agora/common/crypto/Key.d]: Ditto please use final switch here.

File: source/agora/common/crypto/Key.d [hunk 11]
```diff
@@ -320,14 +374,41 @@ public struct Seed
...
+
+    public void toString (scope void delegate(const(char)[]) @safe sink,
+                          PrintMode mode = PrintMode.Obfuscated) const
     {
-        ubyte[1 + Seed.Width + 2] bin;
-        bin[0] = VersionByte.Seed;
-        bin[1 .. $ - 2] = this.data[];
-        bin[$ - 2 .. $] = checksum(bin[0 .. $ - 2]);
-        Base32.encode(bin, sink);
+        switch (mode)
```
Comment (AndrejMitrovic) [inline source/agora/common/crypto/Key.d]: Ditto final switch.

File: source/agora/common/crypto/Key.d [hunk 12]
```diff
@@ -258,6 +260,49 @@ public struct SecretKey
...
+    }
+
+    /***************************************************************************
+
+        Print the secret key
+
+        By default, this function prints the hashed version of the secret key.
+        Changing the mode to `Clear` makes it print the non-hashed value.
+        This mode is intended for debugging. Additionally, no parameterless
+        `toString` is provided for this reason.
```
Comment (AndrejMitrovic) [inline source/agora/common/crypto/Key.d]: This is a very confusing sentence. How about instead:

```suggestion
        This mode is intended for debugging. The parameterless `toString` overload is not implemented, for safety reasons.
```

**Comment by kchulj** [comment]
Any idea why it's failing completely on Mac and Windows?

**Comment by AndrejMitrovic** [comment]
Linker issues, related to dub. There's a way to fix it.

File: source/agora/common/crypto/ECC.d [hunk 13]
```diff
@@ -20,6 +20,7 @@ module agora.common.crypto.ECC;
 import agora.common.Hash;
 import agora.common.Serializer;
 import agora.common.Types;
+import agora.utils.PrettyPrinter;
```
Comment (Geod24) [inline source/agora/common/crypto/ECC.d]: This is your problem. And a big no.

File: source/agora/common/crypto/Key.d [hunk 14]
```diff
@@ -15,8 +15,10 @@ module agora.common.crypto.Key;
 
 import agora.common.crypto.Crc16;
 import agora.common.crypto.ECC;
+import agora.common.Hash;
 import agora.common.Types;
 import agora.common.Serializer;
+import agora.utils.PrettyPrinter;
```
Comment (Geod24) [inline source/agora/common/crypto/Key.d]: So is this

File: source/agora/common/crypto/ECC.d (hunk 13 above)
Comment (AndrejMitrovic) [reply source/agora/common/crypto/ECC.d]: What did @Geod24 tell you yesterday? Always blame drey.

File: source/agora/common/crypto/Key.d [hunk 15]
```diff
@@ -322,7 +321,7 @@ public struct Seed
     }
 
     /// Make sure the sink overload of BitBlob is not picked
-    public void toString (scope void delegate(const(char)[]) sink) const
+    public void toString (scope void delegate(const(char)[]) @safe sink) const
```
Comment (Geod24) [inline source/agora/common/crypto/Key.d]: If you force `@safe` on the delegate, you should force `@safe` on `toString`.

Comment (kchulj) [reply source/agora/common/crypto/Key.d]: Is it ok if `toString` for `Seed` is not `@safe`?
It's using `Base32.encode`, which is not `@safe`.

Comment (Geod24) [reply source/agora/common/crypto/Key.d]: Then why did you change the delegate to be `@safe` ?

Comment (kchulj) [reply source/agora/common/crypto/Key.d]: I was assuming it's possible for only one of them to be `@safe`.. 😅

**Comment by AndrejMitrovic** [comment]
It would be good to also check (for another PR) whether `r` is exposed anywhere. If anyone discovers the `r`, they can derive the private key - and then we're back to having the same problem.

**Comment by AndrejMitrovic** [comment]
`r` being the nonce for the schnorr signatures.

File: source/agora/common/crypto/ECC.d [hunk 16]
```diff
@@ -103,6 +135,20 @@ public struct Scalar
...
+    {
+        Scalar s = Scalar.random();
+        char[] buff1;
+        char[] buff2;
+        scope dg1 = (const(char)[] data) { buff1 ~= data; };
+        scope dg2 = (const(char)[] data) { buff2 ~= data; };
+
+        s.toString(dg1, PrintMode.Obfuscated);
+        s.toString(dg2, PrintMode.Clear);
+        assert(buff1 != buff2);
```
Comment (AndrejMitrovic) [inline source/agora/common/crypto/ECC.d]: This is not really a good test, because it's not actually testing the output, only that it's different. It should test explicitly whether a private key is hidden. So what I suggest is:

- generate a random scalar
- print it out manually
- then hardcode it back into the first line in this test
- assert for both print modes exactly with the expected string output

Comment (kchulj) [reply source/agora/common/crypto/ECC.d]: I updated the `unittest`s. What do you think?

File: source/agora/common/crypto/ECC.d [hunk 17]
```diff
@@ -90,11 +90,43 @@ public struct Scalar
...
+
+    ***************************************************************************/
+
+    public void toString (scope void delegate(const(char)[]) @safe sink,
+                          PrintMode mode = PrintMode.Obfuscated) const @safe
+    {
+        final switch (mode)
+        {
+        case PrintMode.Obfuscated:
+            sink("********");
```
Comment (Geod24) [inline source/agora/common/crypto/ECC.d]:
```suggestion
            sink("**SCALAR**");
```

So if we have multiple obfuscated value, we can tell the types apart.

File: source/agora/common/crypto/ECC.d [hunk 18]
```diff
@@ -103,6 +135,21 @@ public struct Scalar
         return this.data.toString();
```
Comment (Geod24) [inline source/agora/common/crypto/ECC.d]: This is problematic now

File: source/agora/common/crypto/ECC.d (hunk 17 above)
Comment (kchulj) [reply source/agora/common/crypto/ECC.d]: Good idea.

File: source/agora/common/crypto/Key.d [hunk 19]
```diff
@@ -320,14 +423,19 @@ public struct Seed
         return Base32.encode(bin).assumeUnique;
```
Comment (Geod24) [inline source/agora/common/crypto/Key.d]: This needs fixing

File: source/agora/common/crypto/ECC.d (hunk 1 above)
Comment (Geod24) [reply source/agora/common/crypto/ECC.d]: Move this enum to `agora.common.Types` to avoid the duplication

File: source/agora/common/crypto/ECC.d (hunk 18 above)
Comment (kchulj) [reply source/agora/common/crypto/ECC.d]: Yes, and these `string toString ()` methods are used in other `struct`s and `unittest`s.
Would it be a good idea to remove the duplicate and use `scope void delegate` everywhere?

File: source/agora/common/crypto/Key.d (hunk 19 above)
Comment (kchulj) [reply source/agora/common/crypto/Key.d]: Should I remove it, since it's a duplicate of `PrintMode.Clear`?

**Review by Geod24 — APPROVED** [review]
